### PR TITLE
Optimized implementation of `GF2Inverse`

### PR DESCRIPTION
**Pull request by tanujkhattar**
This PR improves the T complexity of `GF2Inverse` from O(m^3) to O(m^2log(m)) following construction from https://arxiv.org/pdf/1209.6348

Updates the tests and adds references.

### CODE REVIEW

File: qualtran/bloqs/gf_arithmetic/gf2_inverse.py [hunk 1]
```diff
@@ -73,15 +75,22 @@ class GF2Inverse(Bloq):
         result: Output RIGHT register of size $m$ that stores $x^{-1}$ from $GF(2^m)$.
         junk: Output RIGHT register of size $m$ and shape ($m - 2$) that stores
             results from intermediate multiplications.
+
+    References:
+        [Efficient quantum circuits for binary elliptic curve arithmetic:
+            reducing T -gate complexity](https://arxiv.org/abs/1209.6348)
+
```
Comment (NoureldinYosri) [inline qualtran/bloqs/gf_arithmetic/gf2_inverse.py:82]:
```suggestion
            section 2.3
```

File: qualtran/bloqs/gf_arithmetic/gf2_inverse.py [hunk 2]
```diff
@@ -96,60 +105,111 @@ def signature(self) -> 'Signature':
...
-
         junk = []
-        for i in range(2, self.bitsize):
-            x = bb.add(GF2Square(self.bitsize), x=x)
-            x, result, new_result = bb.add(GF2Multiplication(self.bitsize), x=x, y=result)
-            junk.append(result)
-            result = new_result
-        x = bb.add(GF2Square(self.bitsize), x=x)
-        return {'x': x, 'result': result} | ({'junk': np.array(junk)} if junk else {})
+        beta = bb.allocate(dtype=self.qgf)
```
Comment (NoureldinYosri) [inline qualtran/bloqs/gf_arithmetic/gf2_inverse.py:140]:
```suggestion
        # The algorithm is descriped on page 4 and 5 of https://arxiv.org/abs/1209.6348 and resembles binary exponentiation.
        # The inverse is computed as (B_{n-1})^2. Where B_1 = x and B_{i+j} = B_i B_j^{2^i}.
        beta = bb.allocate(dtype=self.qgf)
```

File: qualtran/bloqs/gf_arithmetic/gf2_inverse_test.py [hunk 3]
```diff
@@ -53,3 +57,9 @@ def test_gf2_inverse_classical_sim(m):
     bloq = GF2Inverse(m)
     GFM = GF(2**m)
     assert_consistent_classical_action(bloq, x=GFM.elements[1:])
+
```
Comment (NoureldinYosri) [inline qualtran/bloqs/gf_arithmetic/gf2_inverse_test.py:60]: can you add a test for symbolic cost?

**Review by NoureldinYosri — APPROVED** [review]
LGTM%comments

File: qualtran/bloqs/gf_arithmetic/gf2_inverse.py [hunk 4]
```diff
@@ -73,15 +75,22 @@ class GF2Inverse(Bloq):
         result: Output RIGHT register of size $m$ that stores $x^{-1}$ from $GF(2^m)$.
         junk: Output RIGHT register of size $m$ and shape ($m - 2$) that stores
             results from intermediate multiplications.
+
+    References:
+        [Efficient quantum circuits for binary elliptic curve arithmetic:
```
Comment (mpharrigan) [inline qualtran/bloqs/gf_arithmetic/gf2_inverse.py:80]: Can you keep these links on one line? https://qualtran.readthedocs.io/en/latest/Autodoc.html#references

File: qualtran/bloqs/gf_arithmetic/gf2_inverse.py (hunk 2 above)
Comment (mpharrigan) [reply qualtran/bloqs/gf_arithmetic/gf2_inverse.py:140]: cool, yeah. I'd see if this could make its way to the class docstring so it shows up in the docs

**Review by mpharrigan — APPROVED** [review]
lgtm modulo nits and nour's nits

File: qualtran/bloqs/gf_arithmetic/gf2_inverse_test.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/bloqs/gf_arithmetic/gf2_inverse_test.py:60]: The test for symbolic cost exists above in `test_gf2_inverse_symbolic_toffoli_complexity`

**Comment by tanujkhattar** [comment]
Addressed all nits. Merging now

File: qualtran/bloqs/gf_arithmetic/gf2_inverse.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/gf_arithmetic/gf2_inverse.py:141]: Added to the class docstring and regenerated the docs

**Pull request merged**